test/e2e/gmsa.go

```diff
@@ -0,0 +1,201 @@
+//go:build e2e
+// +build e2e
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/2020-09-01/compute/mgmt/compute"
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	capz "sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func configureGmsa(ctx context.Context, workloadProxy, bootstrapClusterProxy framework.ClusterProxy, namespace, clusterName string, config *clusterctl.E2EConfig) {
+	settings, err := auth.GetSettingsFromEnvironment()
+	authorizer, err := settings.GetAuthorizer()
+	Expect(err).NotTo(HaveOccurred())
+	subId := settings.GetSubscriptionID()
+
+	Expect(err).NotTo(HaveOccurred())
+	keyVaultClient := keyvault.New()
+
+	vmClient := compute.NewVirtualMachinesClient(subId)
+	vmClient.Authorizer = authorizer
+
+	networkClient := network.NewVirtualNetworkPeeringsClient(subId)
+	networkClient.Authorizer = authorizer
+
+	//override to use keyvault management endpoint
+	settings.Values[auth.Resource] = fmt.Sprintf("%s%s", "https://", azure.PublicCloud.KeyVaultDNSSuffix)
+	keyvaultAuthorizer, err := settings.GetAuthorizer()
+	keyVaultClient.Authorizer = keyvaultAuthorizer
+
+	// Wait for the Cluster nodes to be ready (this is different than capi's ready as cni needs to finish initializing)
+	windowsCalico := &appsv1.DaemonSet{
```

Comment:
this test is assuming calico CNI, would we ever want it run with other CNIs? Would it be equivalent to check the nodes are all "Ready"?

Reply:
I did this before I learned about: `ControlPlaneWaiters: clusterctl.ControlPlaneWaiters{`  Maybe that would be a better approach? It might help with some flakes that happen before the Windows pods are fully ready too.